rust/src/mqtt/parser.rs

```diff
@@ -0,0 +1,631 @@
+           >> flags: parse_connect_variable_flags
+           >> keepalive: be_u16
+           >> properties: call!(parse_properties, protocol_version == 5)
+           >> client_id: parse_mqtt_string
+           >> will_properties: call!(parse_properties, protocol_version == 5 && flags.4 != 0)
+           >> will_topic: cond!(flags.4 != 0, parse_mqtt_string)
+           >> will_message: cond!(flags.4 != 0, parse_mqtt_binary_data)
+           >> username: cond!(flags.0 != 0, parse_mqtt_string)
+           >> password: cond!(flags.1 != 0, parse_mqtt_binary_data)
+           >>  (
+                 MQTTConnectData {
+                   protocol_string: protocol_string,
+                   protocol_version: protocol_version,
+                   username_flag: flags.0 != 0,
+                   password_flag: flags.1 != 0,
+                   will_retain: flags.2 != 0,
+                   will_qos: flags.3 as u8,
+                   will_flag: flags.4  != 0,
+                   clean_session: flags.5 != 0,
+                   keepalive: keepalive,
+                   client_id: client_id,
+                   will_topic: will_topic,
+                   will_message: will_message,
+                   username: username,
+                   password: password,
+                   properties: properties,
+                   will_properties: will_properties,
+                 }
+               )
+       ));
+
+named_args!(pub parse_connack(protocol_version: u8)<MQTTConnackData>,
+       do_parse!(
+           topic_name_compression_response: be_u8
+           >> retcode: be_u8
+           >> properties: call!(parse_properties, protocol_version == 5)
+           >>  (
+                 MQTTConnackData {
+                   session_present: (topic_name_compression_response & 1) != 0,
+                   return_code: retcode,
+                   properties: properties,
+                 }
+               )
+       ));
+
+named_args!(pub parse_publish(protocol_version: u8, has_id: bool)<MQTTPublishData>,
+       do_parse!(
+           topic: parse_mqtt_string
+           >> message_id: cond!(has_id, be_u16)
+           >> properties: call!(parse_properties, protocol_version == 5)
+           >> message: rest
+           >>  (
+                 MQTTPublishData {
+                   topic: topic,
+                   message_id: message_id,
+                   message: message.to_vec(),
+                   properties: properties,
+                 }
+               )
+       ));
+
+#[inline]
+fn parse_msgidonly(
+    input: &[u8],
+    mut remaining_len: usize,
+    protocol_version: u8,
+) -> IResult<&[u8], MQTTMessageIdOnly> {
+    if protocol_version < 5 {
+        // before v5 we don't even have to care about reason codes
+        // and properties, lucky us
+        return parse_msgidonly_v3(input);
+    }
+    match be_u16(input) {
+        Ok((rem, message_id)) => {
+            remaining_len -= 2;
+            if remaining_len == 0 {
```

Comment:
This feels dangerous and of integer underflow

Can we not have this with nom ?
like `alt! (complete! (be_u8) | value!(0))` for reason_code

Reply:
Well, we need to check here whether the message has omitted reason code and properties, and need to break early and not try parsing the properties then. I don't know how to do that with nom macros, it's a similar problem to the one you pointed out earlier. TBH I'd rather have explicit, documented parsing steps here just to make it clear what is happening, but I agree that there should be extra checks to make sure there is no integer underflow in these unchecked arithmetics. Will add them.

Reply:
Does not my nom `alt! (complete! (be_u8) | value!(0))` do the trick for `reason_code` ?
Then for `properties`, you can wrap with `alt! (complete! ( STUFF ))` to get `None ` in the relevant case, or the properties if you have data for them.

About the general philosophy of explicit documented parsing steps versus nom magic, this is a good question I do not have the answer to.
It seemed to me that there was more nom magic than explicit parsing steps, so I tried this approach in my parsers even if it felt not natural to me.